internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+	ctx := context.Background()
+	bmh := &bmh_v1alpha1.BareMetalHost{}
+
+	if err := r.Get(ctx, req.NamespacedName, bmh); err != nil {
+		return reconcileComplete{}.Result()
+	}
+
+	// Let's reconcile the BMH
+	result := r.reconcileBMH(bmh)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, bmh)
+		if err != nil {
+			r.Log.Errorf("Error updating hardwaredetails %v", err)
+			return reconcileError{err}.Result()
+		}
+
+		return result.Result()
+	}
+
+	// Should we check the status of the BMH resource?
+	// Provisioning / Deprovisioning ?
+	// What happens if we do the reconcile on a BMH that
+	// is in a Deprovisioning state?
+
+	agent := r.findAgent(ctx, bmh)
+
+	// handle multiple agents matching the
+	// same BMH's Mac Address
+
+	if agent == nil {
+		return result.Result()
+	}
+
+	// if we get to this point, it means we have found an Agent that is associated
+	// with the BMH being reconciled. We will call both, reconcileAgentSpec and
+	// reconcileAgentInventory, every time. The logic to decide whether there's
+	// any action to take is implemented in each functoin respectively.
+	result = r.reconcileAgentSpec(bmh, agent)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, agent)
+		if err != nil {
+			r.Log.Errorf("Error updating agent %v", err)
+			return reconcileError{err}.Result()
+		}
+
+		return result.Result()
+	}
+
+	result = r.reconcileAgentInventory(bmh, agent)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, bmh)
+		if err != nil {
+			r.Log.Errorf("Error updating hardwaredetails %v", err)
+			return reconcileError{err}.Result()
+		}
+	}
+
+	return result.Result()
+}
+
+// This reconcile step takes care of copying data from the BMH into the Agent.
+//
+// The Agent's spec reconcile will only happen for non-Approved agents. These are
+// considered to be `new` and, at this point, the agent has not started the deployment
+// yet.
+//
+// We are only interested in the following data:
+//
+// - Agent role: bmac.adi.openshift.io/role
+// - Hostname: bmac.adi.openshift.io/hostname
+// - Machine Config Pool: bmac.adi.openshift.io/machine-config-pool
+//
+// Unless there are errors, the agent should be `Approved` at the end of this
+// reconcile and a label should be set on it referencing the BMH. No changes to
+// the BMH should happen in this reconcile step.
+func (r *BMACReconciler) reconcileAgentSpec(bmh *bmh_v1alpha1.BareMetalHost, agent *adiiov1alpha1.Agent) reconcileResult {
+
+	if agent.Spec.Approved {
+		return reconcileComplete{}
+	}
+
+	r.Log.Debugf("Started Agent Spec reconcile for agent %s/%s and bmh %s/%s", agent.Namespace, agent.Name, bmh.Namespace, bmh.Name)
+
+	// Do all the copying from the BMH annotations to the agent.
+
+	annotations := bmh.ObjectMeta.GetAnnotations()
+	if val, ok := annotations[BMH_AGENT_ROLE]; ok {
+		switch val {
+		case "master":
+			agent.Spec.Role = models.HostRoleMaster
+		case "worker":
+			agent.Spec.Role = models.HostRoleWorker
+			// should we do something?
```

Comment:
Maybe just copy the value directly instead of doing validation here? I'm not sure validating at this point is helpful.

Reply:
> Maybe just copy the value directly instead of doing validation here? I'm not sure validating at this point is helpful.

A conversation with @avishayt suggested that we only want to handle `master`/`worker` in this scenario.

Can we confirm if this is the case?

Reply:
I guess what @mhrivnak is suggesting is that the value will be validated by the Agent controller once you set it, and if it looks the same to the user then you may as well leave it out so it doesn't need to be maintained if we add more values in the future.

Reply:
(Similar to how you're not checking that hostname is valid below, but rather relying on the Agent controller)

Reply:
@mhrivnak I agree with you, hence the comment I left there.

To be honest, it's still a bit unclear to me how we want to manage validations and errors in this controller. For now, we are just reconciling when some errors occur.

To me, the validation should happen on the CR side. If one value is not set properly, then the resource's reconcile should find the issue, and set the validation error. I think it would be ideal to keep BMAC as the controller that copies data around, without handling resource-specific validations. This may not always be possible, tho! 

All the above to say that I'll switch this to just set the role as-is in the agent and let the agent take care of the rest.

Reply:
@avishayt  I just read your comment, I had written the reply to @mhrivnak before noticing that you had reviewed the PR too. I'm glad we all agree :)